syntax = "proto3";

package received_item;

import "google/protobuf/empty.proto";
import "logistics/protos/inventory_proto/inventory.proto";

service ReceivedItemController {
    rpc List(ReceivedItemListRequest) returns (stream ReceivedItem) {}
    rpc Create(ReceivedItem) returns (ReceivedItem) {}
    rpc Retrieve(ReceivedItemRetrieveRequest) returns (ReceivedItem) {}
    rpc Update(ReceivedItem) returns (ReceivedItem) {}
    rpc Destroy(ReceivedItem) returns (google.protobuf.Empty) {}

    rpc Make(MakeReceivedItemRequest) returns (stream ReceivedItem) {}
}

enum ReceivedItemStatus {
    SOURCING_REQUIRED = 0;
    ON_RECEIVING = 1;
    RECEIVED = 2;
    OUT_OF_STOCK = 3;
    CANCELED = 4;
}

message ReceivedItem {
    int64 id = 1;
    int64 order_id = 2;
    int64 order_item_id = 3;
    string code = 4;
    ReceivedItemStatus status = 5;
    string product_id = 6;
    string product_name = 7;
    string product_img = 8;
    string brand_korname = 9;
    string brand_keyname = 10;
    string size = 11;
    optional string color = 12;
    string created_at = 13;
    string updated_at = 14;
    optional string deleted_at = 15;
    optional inventory.Inventory inventory = 16;
}

message ReceivedItemListRequest {
    optional string order_id = 1;
    optional string order_item_id = 2;
    optional string product_name = 3;
    optional string brand_korname = 4;
    optional string brand_keyname = 5;
    optional string code = 6;
    repeated string statuses = 7;
    map<string, int64> query_params = 8;
}

message ReceivedItemListResponse {
    int64 count = 1;
    string next = 2;
    string previous = 3;
    repeated ReceivedItem results = 4;
}

message ReceivedItemRetrieveRequest {
    int64 id = 1;
}

message MakeReceivedItemRequest {
    int64 order_id = 1;
    int64 order_item_id = 2;
    string order_item_code = 3;
    string brand_korname = 4;
    string brand_keyname = 5;
    string product_id = 6;
    string product_img = 7;
    string product_name = 8;
    string size = 9;
    optional string color = 10;
    int64 item_quantity = 11;
    int64 request_quantity = 12;
    bool force = 13;
}
